cmd/gui/gui.go

```diff
@@ -0,0 +1,154 @@
+package gui
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	log "github.com/cihub/seelog"
+	"github.com/gorilla/mux"
+)
+
+var (
+	apiKey   string
+	listener net.Listener
+)
+
+// Message struct is for the JSON messages received from a client POST request
+type Message struct {
+	ReqType string `json:"req_type"`
+	Data    string `json:"data"`
+	Payload string `json:"payload"`
+}
+
+// StopGUIServer closes the connection to the HTTP server
+func StopGUIServer() {
+	if listener != nil {
+		listener.Close()
+	}
+}
+
+// StartGUIServer creates the router and starts the HTTP server
+func StartGUIServer() error {
+	apiKey = config.Datadog.GetString("api_key")
+	port := ":" + config.Datadog.GetString("GUI_port")
+
+	if config.Datadog.GetString("GUI_port") == "-1" {
+		log.Infof("Port -1 specified: not starting the GUI server.")
+		return nil
+	}
+
+	// Instantiate the gorilla/mux router
+	router := mux.NewRouter()
+
+	// Serve the (secured) index page on the default endpoint
+	router.Handle("/", accessAuth(http.FileServer(http.Dir("view/private/"))))
+
+	// Mount our public filesystem at the view/{path} route
+	router.PathPrefix("/view/").Handler(http.StripPrefix("/view/", http.FileServer(http.Dir("view/public/"))))
+
+	// Mount our secured filesystem at the private/{path} route
+	router.PathPrefix("/private/").Handler(http.StripPrefix("/private/", accessAuth(http.FileServer(http.Dir("view/private/")))))
+
+	// Handle requests from clients
+	router.Handle("/req", authenticate(http.HandlerFunc(handler))).Methods("POST")
+
+	listener, e := net.Listen("tcp", port)
```

Comment:
The server should definitely listen only on localhost `127.0.0.1`, not all interfaces (`0.0.0.0`)

Reply:
It needs to run on 0.0.0.0 in case the user is running the agent on a VM or in a container. The fact that the interface is open to the whole local network was the original reason behind adding API key authentication (whether or not that is actually sufficient is another question though)

Reply:
So, I'll approve as is so we can get this merged in.  But this is going to need to be re-thought.  API key auth isn't sufficient (for a lot of reasons), and open to everyone isn't going to fly.